packages/block-editor/src/components/block-list/index.js

```diff
@@ -90,59 +109,40 @@ function Items( {
 	__experimentalAppenderTagName,
 	__experimentalLayout: layout = defaultLayout,
 } ) {
-	const [ intersectingBlocks, setIntersectingBlocks ] = useState( new Set() );
-	const intersectionObserver = useMemo( () => {
-		const { IntersectionObserver: Observer } = window;
-
-		if ( ! Observer ) {
-			return;
-		}
-
-		return new Observer( ( entries ) => {
-			setIntersectingBlocks( ( oldIntersectingBlocks ) => {
-				const newIntersectingBlocks = new Set( oldIntersectingBlocks );
-				for ( const entry of entries ) {
-					const clientId = entry.target.getAttribute( 'data-block' );
-					const action = entry.isIntersecting ? 'add' : 'delete';
-					newIntersectingBlocks[ action ]( clientId );
-				}
-				return newIntersectingBlocks;
-			} );
-		} );
-	}, [ setIntersectingBlocks ] );
-	const { order, selectedBlocks } = useSelect(
+	const { order, selectedBlocks, visibleBlocks } = useSelect(
 		( select ) => {
-			const { getBlockOrder, getSelectedBlockClientIds } = select(
-				blockEditorStore
-			);
+			const {
+				getBlockOrder,
+				getSelectedBlockClientIds,
+				__unstableGetVisibleBlocks,
+			} = select( blockEditorStore );
 			return {
 				order: getBlockOrder( rootClientId ),
 				selectedBlocks: getSelectedBlockClientIds(),
+				visibleBlocks: __unstableGetVisibleBlocks(),
```

Comment:
There could be a possible optimization opportunity here: `visibleBlocks` cares only about blocks that are _top-level_, because that's where async mode is being set. Nested blocks' visibility is redundant. But `__unstableGetVisibleBlocks` will return all visible blocks, causing wasted rerenders. Happens commonly when I, e.g., scroll away only a part of a complex site header structure. The top-level block remains visible, but some deeply nested blocks become invisible:

<img width="1129" alt="Screenshot 2022-09-12 at 13 16 06" src="https://user-images.githubusercontent.com/664258/189640384-f0f8b901-84d9-471b-8473-5a65acc5b5ca.png">

Not sure how impactful this really is -- just noting down something I observed when working on https://github.com/WordPress/gutenberg/pull/42525#issuecomment-1243557542.

Reply:
In that case I think it doesn't work. Would need to verify carefully to be 100% sure, but only the top-level `BlockListItems` component (in `@wordpress/block-editor`) ever renders `<AsyncModeProvider>`, switching top-level blocks between sync and async as they become visible or invisible.

If there is a group block with many paragraphs, then these paragraph blocks have correct `isBlockVisible` state in the store, but the sync/async mode switching doesn't happen, because they're nested.

Reply:
AsyncModeProvider is rendered around each block https://github.com/WordPress/gutenberg/blob/c1cf175c962ac2239020b80fb6a02341f7205f0b/packages/block-editor/src/components/block-list/index.js#L130

Reply:
Turns out I misunderstood how nested blocks are rendered! I thought that the `BlockListItems` component with all the async providers is rendered only at the root, but it's used by every block that supports nested blocks. That ensures that the sync/async switching works at all levels.

Still, `__unstableGetVisibleBlocks` could be optimized a bit? Instead of returning all visible blocks, it could take a `rootClientId` argument and return only immediate children of a given parent block. Because now, whenever there's a visibility change at any level, all `BlockListItems` instances are rerendered. And there are 24 instances when I edit the main template in Twenty Twenty-Two. With `__unstableGetVisibleBlocks( rootClientId )`, only the one affected `BlockListItems` instance would rerender.

Reply:
>  all BlockListItems instances are rerendered. And there are 24 instances when I edit the main template in Twenty Twenty-Two. With __unstableGetVisibleBlocks( rootClientId ), only the one affected BlockListItems instance would rerender.

Yes, that sounds like something we could try. That said, It's not always obvious what's more performant: a more complex selector or a loop that does not much (BlockListBlock components are memoized).